compiler/src/compiler/cst_to_ast.rs

```diff
@@ -140,26 +144,98 @@ impl LoweringContext {
-                assert!(
-                    matches!(opening_quote.kind, CstKind::DoubleQuote),
-                    "Text needs to start with opening double quote, but started with {}.",
-                    opening_quote
-                );
+                let opening_single_quote_count = match &opening_quote.kind {
+                    CstKind::OpeningText {
+                        opening_single_quotes,
+                        opening_double_quote: box Cst {
+                            kind: CstKind::DoubleQuote,
+                            ..
+                        }
+                    } if opening_single_quotes
+                        .iter()
+                        .all(|opening_single_quote| -> bool {
+                            matches!(opening_single_quote.kind, CstKind::SingleQuote)
+                        }) => opening_single_quotes.len(),
+                    _ => panic!("Text needs to start with any number of opening single quotes followed by an opening double quote, but started with {}.", opening_quote)
+                };
 
-                let text = parts
-                    .iter()
-                    .filter_map(|it| match it {
+                let mut lowered_parts = vec![];
+                for part in parts {
+                    match part {
                         Cst {
                             kind: CstKind::TextPart(text),
                             ..
-                        } => Some(text),
+                        } => {
+                            // TODO: Combine successive text parts during compile time
+                            let string = self.create_string_without_id_mapping(text.clone());
+                            let text_part =
+                                self.create_ast(cst.id, AstKind::TextPart(TextPart(string)));
+                            lowered_parts.push(text_part);
+                        },
+                        Cst {
+                            kind: CstKind::TextPlaceholder { 
+                                opening_curly_braces,
+                                expression,
+                                closing_curly_braces
+                            },
+                            ..
+                        } => {
+                            if opening_curly_braces.len() != (opening_single_quote_count + 1)
+                                || opening_curly_braces
+                                    .iter()
+                                    .all(|opening_curly_brace| !matches!(opening_curly_brace.kind, CstKind::OpeningCurlyBrace)) 
+                            {
+                                panic!(
+                                    "Text placeholder needs to start with {} opening curly braces, but started with {}.", 
+                                    opening_single_quote_count + 1, 
+                                    opening_curly_braces.iter().map(|cst| format!("{}", cst)).join("")
+                                )
```

Comment:
I'm wondering: Can you interpolate closures? As in `"{{foo}}"` being equivalent to this:

```candy
closure = { foo }
"{closure}"
```

Reply:
Yes this works. If your text starts with _n_ single quotes, the first _n+1_ curly braces will be parsed as the start of an interpolation. Any more curly braces will be parsed as part of the expression.

Unfortunately, this means that text can't contain curly braces directly before or after an interpolation. Backtracking could be used to find out if additional curly braces should be part of the text or the expression, but I think this would result in an unpredictable syntax. A better solution might be the addition of other special characters before and after the interpolation.

Reply:
A bug related to this gets fixed in [1abf946](https://github.com/candy-lang/candy/pull/258/commits/1abf94612f42a2cd0b336a3241586540c322e82d).

Reply:
Ohhh, you're right. I didn't think about braces that are part of the text itself. Basically, `"{{foo}}"` could get parsed as either `"{<closure returning foo>}"` or `<closure returning a closure returning foo>`.

Tbh, we probably don't often interpolate closures, so I think it would be even fine to require spaces in that case. But I agree that we need to think more about this and any handling is okay for now.

Reply:
I changed the parser to require a space before closures in 4803279.